cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -1282,6 +1292,251 @@ def _check_column_availability(self):
+            self.errors.append(
+                "uns['spatial'] is required for obs['assay_ontology_term_id'] values "
+                "'EFO:0010961' (Visium Spatial Gene Expression) and 'EFO:0030062' (Slide-seqV2)."
+            )
+            return
+
+        # is_single is required.
+        uns_spatial = self.adata.uns["spatial"]
+        if "is_single" not in uns_spatial:
+            self.errors.append("uns['spatial'] must contain the key 'is_single'.")
+            # Exit if is_single is missing as all further checks are dependent on its value.
+            return
+
+        # is_single must be a boolean.
+        uns_is_single = uns_spatial["is_single"]
+        if not isinstance(uns_is_single, np.bool_):
+            self.errors.append(f"uns['spatial']['is_single'] must be of boolean type, it is {type(uns_is_single)}.")
+            # Exit if is_single is not valid as all further checks are dependent on its value.
+            return
+
+        # Check there is at most one library_id.
+        uns_spatial_keys = list(uns_spatial.keys())
+        library_ids = list(filter(lambda x: x != "is_single", uns_spatial_keys))
+        if len(library_ids) > 1:
+            self.errors.append("uns['spatial'] must contain only one library_id.")
+            # Exit if there is more than one library_id as we don't know which library_id to validate.
+            return
+
+        # library_id is forbidden if assay is not Visium or is_single is false.
+        is_visium = self._is_visium()
+        is_visium_and_uns_is_single = is_visium and uns_is_single
+        if len(library_ids) > 0 and not is_visium_and_uns_is_single:
+            self.errors.append(
+                "uns['spatial'][library_id] is only allowed for obs['assay_ontology_term_id'] "
+                "'EFO:0010961' (Visium Spatial Gene Expression) and uns['spatial']['is_single'] is True."
+            )
+            # Exit as library_id is not allowed.
+            return
+
+        # Exit if we're not dealing with Visium and _is_single True as no further checks are necessary.
+        if not is_visium_and_uns_is_single:
+            return
+
+        # library_id is required if assay is Visium and is_single is True.
+        if len(library_ids) == 0:
+            self.errors.append(
+                "uns['spatial'] must contain the key 'library_id' for obs['assay_ontology_term_id'] "
+                "'EFO:0010961' (Visium Spatial Gene Expression) and uns['spatial']['is_single'] is True."
+            )
+            # Exit as library_id is missing.
+            return
+
+        # Confirm shape of library_id is valid: allowed keys are images and scalefactors.
+        library_id_key = library_ids[0]
+        uns_library_id = uns_spatial[library_id_key]
+        if not self._has_no_extra_keys(uns_library_id, ["images", "scalefactors"]):
+            self.errors.append("uns['spatial'][library_id] can only contain the keys 'images' and 'scalefactors'.")
```

Comment:
nit: now that we have validated there is only one library_id, we can substitute in the key we're referencing (here and in the subsequent error messages below)

i.e. `f"uns['spatial']['{uns_library_id}'] can only contain the keys 'images' and 'scalefactors'.""`

this more of a nice-to-have, but pointing it out since there are a couple of other requested changes